File: app/src/main/java/com/desarollo/luisvillalobos/gardenkit/View/FormDevice.java
package com.desarollo.luisvillalobos.gardenkit.View;

import android.content.Context;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.EditText;

import com.desarollo.luisvillalobos.gardenkit.Controller.SetUpActivity;
import com.desarollo.luisvillalobos.gardenkit.Controller.DatabaseAccess;
import com.desarollo.luisvillalobos.gardenkit.R;

public class FormDevice extends AppCompatActivity {

    /*
    private ImageButton imgBtnBack;
    private EditText txtDescription;
    private EditText txtApiKey;
    private EditText txtDevice;
    private EditText txtUser;
    private Button btnAdd;
*/

    private EditText x;

    private Context context;
    private DatabaseAccess databaseAccess;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.form_device);
        context = getBaseContext();

        //Configuración de Activity
        SetUpActivity.hiderActionBar(this);
        SetUpActivity.hideStatusBar(this);
        SetUpActivity.hideSoftKeyboard(this);

        x = (EditText) findViewById(R.id.in_apikey);


/*
        //Instanciando los Views
        imgBtnBack = (ImageButton) findViewById(R.id.btnBack);
        txtDescription = (EditText) findViewById(R.id.txtDescription);
        txtApiKey = (EditText) findViewById(R.id.txtApiKey);
        txtDevice = (EditText) findViewById(R.id.txtDevice);
        txtUser = (EditText) findViewById(R.id.txtUser);
        btnAdd = (Button) findViewById(R.id.btnAdd);

        //Configurando las views
        btnAdd.setOnClickListener(new AddBtnClick());
        imgBtnBack.setOnClickListener(new BackImgBtnClick());
        */
    }
/*
    class AddBtnClick implements View.OnClickListener {

        @Override
        public void onClick(View view) {
            Intent returnIntent = getIntent();
            btnAdd.setEnabled(true);
            btnAdd.setClickable(true);
            if (txtDescription.getText().toString().trim().length() != 0 && txtApiKey.getText().toString().trim().length() != 0 && txtDevice.getText().toString().trim().length() != 0 && txtUser.getText().toString().trim().length() != 0) {
                Device device = new Device(txtDescription.getText().toString().trim(), txtApiKey.getText().toString().trim(), txtDevice.getText().toString().trim(), txtUser.getText().toString().trim());
                returnIntent.putExtra("object", device);
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
                setResult(Activity.RESULT_OK, returnIntent);
                finish();
            } else {
                //overridePendingTransition( 0, 0);
                Toast.makeText(context, "No se rellenaron los campos correctamentre", Toast.LENGTH_SHORT).show();
            }
            btnAdd.setEnabled(true);
            btnAdd.setClickable(true);
        }
    }

    class BackImgBtnClick implements View.OnClickListener {

        @Override
        public void onClick(View view) {
            Intent returnIntent = getIntent();
            if (txtDescription.getText().toString().trim().length() != 0 && txtApiKey.getText().toString().trim().length() != 0 && txtDevice.getText().toString().trim().length() != 0 && txtUser.getText().toString().trim().length() != 0) {
                Device device = new Device(txtDescription.getText().toString(), txtApiKey.getText().toString(), txtDevice.getText().toString(), txtUser.getText().toString());
                returnIntent.putExtra("object", device);
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
                setResult(Activity.RESULT_OK, returnIntent);
                finish();
            } else {
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                returnIntent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
                setResult(Activity.RESULT_CANCELED, returnIntent);

                finishFromChild(getParent());
                //finish();
            }
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            //Intent intent = new Intent(context, ListDevices.class);
            //intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            //intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
            //startActivity(intent);
            Intent returnIntent = getIntent();
            returnIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            returnIntent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
            setResult(Activity.RESULT_CANCELED, returnIntent);
            //finishFromChild(getParent());
            finish();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }
    */
}

File: app/src/main/java/com/desarollo/luisvillalobos/gardenkit/Activity/FormDevice.kt
package com.desarollo.luisvillalobos.gardenkit.Activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.opengl.Visibility
import android.os.Bundle
import android.os.Parcelable
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.InputType
import android.util.Log
import android.view.View
import android.widget.TextView
import android.widget.Toast
import com.desarollo.luisvillalobos.gardenkit.Controller.DBHelper
import com.desarollo.luisvillalobos.gardenkit.Controller.SetUpActivity
import com.desarollo.luisvillalobos.gardenkit.Model.Device
import com.desarollo.luisvillalobos.gardenkit.R
import kotlinx.android.synthetic.main.form_device.*
import android.widget.EditText



class FormDevice : AppCompatActivity(), View.OnClickListener {

    private var key: Int = 0
    private lateinit var returnIntent: Intent
    private var device: Device? = null
    private var requestOption: Int = 0
    private var idDevice: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.form_device)
        setup()
    }

    private fun setup() {
        //Screen
        SetUpActivity.hiderActionBar(this)
        SetUpActivity.hideStatusBar(this)
        SetUpActivity.hideSoftKeyboard(this)

        //Listeners
        btn_action.setOnClickListener(this)
        btn_home.setOnClickListener(this)

        //Extras
        returnIntent = intent
        returnIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NO_HISTORY
        key = Integer.parseInt("" + getSharedPreferences(Login.PREFS_NAME, Context.MODE_PRIVATE).getString(Login.USERID, null))

        //GetExtras
        device = intent.getParcelableExtra(ListDevices.DEVICE_PARCEABLE_TAG)
        requestOption = intent.extras!!.getInt(ListDevices.DEVICE_OPTION_TAG)
        idDevice = intent.extras!!.getInt(ListDevices.DEVICE_ID_TAG)

        when (requestOption) {
            ListDevices.REQUEST_FORMDEVICE_READ -> {
                in_name.setText(device?.name)
                in_device.setText(device?.device_request)
                in_apikey.setText(device?.apikey_request)
                in_description.setText(device?.description)

                in_name.inputType = InputType.TYPE_NULL
                in_device.inputType = InputType.TYPE_NULL
                in_apikey.inputType = InputType.TYPE_NULL
                in_description.inputType = InputType.TYPE_NULL
                btn_action.visibility = View.INVISIBLE
            }
            ListDevices.REQUEST_FORMDEVICE_UPDATE -> {
                in_name.setText(device?.name)
                in_device.setText(device?.device_request)
                in_apikey.setText(device?.apikey_request)
                in_description.setText(device?.description)
                btn_action.setText("Actualizar dispositivo")
            }
        }
    }

    //ClickListeners buttons implementation
    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.btn_action -> actionBtnClick()
            R.id.btn_home -> homeBtnClick()
        }
    }

    private fun actionBtnClick() {
        if (!fieldsNotNull()) {
            toast("Verifique los datos")
            return
        }

        when (requestOption) {
            ListDevices.REQUEST_FORMDEVICE_ADD -> {
                val deviceAux = Device(in_name.text.toString(),
                        in_device.text.toString(),
                        in_apikey.text.toString(),
                        in_description.text.toString(),
                        key)
                returnIntent.putExtra(ListDevices.DEVICE_PARCEABLE_TAG, deviceAux)
                setResult(Activity.RESULT_OK, returnIntent)
                finish()
            }
            ListDevices.REQUEST_FORMDEVICE_UPDATE -> {
                val deviceAux = Device(in_name.text.toString(),
                        in_device.text.toString(),
                        in_apikey.text.toString(),
                        in_description.text.toString(),
                        key)
                returnIntent.putExtra(ListDevices.DEVICE_PARCEABLE_TAG, deviceAux)
                returnIntent.putExtra(ListDevices.DEVICE_ID_TAG, idDevice)
                setResult(Activity.RESULT_OK, returnIntent)
                finish()
            }
        }
    }

    private fun homeBtnClick() {
        setResult(Activity.RESULT_CANCELED, returnIntent)
        finishFromChild(parent)
    }

    //Extras
    private fun fieldsNotNull(): Boolean {
        return in_apikey.text.isNotEmpty() &&
                in_description.text.isNotEmpty() &&
                in_device.text.isNotEmpty() &&
                in_name.text.isNotEmpty()
    }

    private fun disableEditText(editText: EditText) {
        editText.isFocusable = false
        //editText.isEnabled = false
        editText.isCursorVisible = false
        editText.keyListener = null
        editText.setBackgroundColor(Color.TRANSPARENT)
    }

    //Message
    private fun toast(message: String) {
        Toast.makeText(baseContext, "$message", Toast.LENGTH_LONG).show()
    }

    private fun log(message: String) {
        Log.e("BitZero", "$message")
    }
}